### PR TITLE
Define a platform hook interface and cleanup misuse of private calls

### PR DESCRIPTION
**Pull request by nashif**
An attempt to cleanup misuse of private API call to define platform hooks and generalize the definition of hooks implemented by platforms overriding architecture implementation.

This also generalizes some hook implementations done for some architectures and made such implementation arch agnostic, allowing hooks to be implemented by any architecture.

Relates to https://github.com/zephyrproject-rtos/zephyr/issues/58007

### CODE REVIEW

**Review by gmarull — COMMENTED** [review]
changes make sense to me. we should document reserved internal prefixes like arch, platform, etc. Some CI checks would also be nice to prevent misuses (e.g. forbid `arch_`/`platform_` calls in samples/ or drivers/)

**Review by galak — COMMENTED** [review]
any reason this can't be done with weak functions?

**Comment by nashif** [comment]
> any reason this can't be done with weak functions?

This has been the approch for some interfaces. This PR is mostly changing the naming and moving a fewthings upo to the architecture level, but not changing how the interfaces are implemented.
We had a few change proposals floating around removing usage of _weak given it is considered not stanard but we still have many usages in the tree. Whether we use _weak or some other mechanism is something we need decide on separately I guess.

File: doc/hardware/porting/custom.rst [hunk 1]
```diff
@@ -0,0 +1,16 @@
+.. _platform_cusomizations:
+
+Hardware with Custom Interfaces
```
Comment (mbolivar-ampere) [inline doc/hardware/porting/custom.rst:3]: Seems like a good first step towards closing the documentation gap between "board porting" and "soc" or "arch" porting.

**Comment by amr-sc** [comment]
> > > any reason this can't be done with weak functions?
> > 
> > 
> > This has been the approch for some interfaces. This PR is mostly changing the naming and moving a fewthings upo to the architecture level, but not changing how the interfaces are implemented. We had a few change proposals floating around removing usage of _weak given it is considered not stanard but we still have many usages in the tree. Whether we use _weak or some other mechanism is something we need decide on separately I guess.
> 
> Probably more useful than having weak functions would be to fail at compile time if a `CUSTOM` option has been selected and the prescribed interface has not been implemented.
> 
> Is there a particular use for weak functions in this case?
> 
> The other bit of input I had here was that there are systems where there are multiple (system) interrupt controllers.

Hi @cfriedt , I think the _weak functions could be useful to override the default arch_ level implementations (which follow the Spec) in case some vendors doesn't conform to the Spec, or the Spec is ambiguous. I consider it more like a sign of a temporary workaround, or the last resort solution to plug-in some non-standard implementation.

**Comment by amr-sc** [comment]
> > > > > any reason this can't be done with weak functions?
> 
> > > Is there a particular use for weak functions in this case?
> 
> > Hi @cfriedt , I think the _weak functions could be useful to override the default arch_ level implementations (which follow the Spec) in case some vendors doesn't conform to the Spec, or the Spec is ambiguous.
> 
> Sure - it's just that weak symbols are not part of ISO C or ISO C++ and are very much ELF specific.
> 
> We can achieve the same thing with a well-defined interface and having a Kconfig option to choose between the default (arch) and overridden (platform) solution.

Oh, I've got your point.
The explicit policy with default arch_ and overridden platform_ API is for sure better than the _weak approach.
Thanks! 

P.S.: looks like this may provide a guideline for #65824 as well

File: doc/hardware/porting/custom.rst [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+###############################
+
+When the provided support in Zephyr does not cover certain hardware platforms,
+some interfaces can be implemented directly by the platform or a custom
+architecture interface can be used and resulting in more flexibility and
+hardware support of differebt variants of the same hardware.
```
Comment (aescolar) [inline doc/hardware/porting/custom.rst:9]: typo differebt

File: doc/hardware/porting/custom.rst [hunk 3]
```diff
@@ -0,0 +1,16 @@
+.. _platform_cusomizations:
+
+Hardware with Custom Interfaces
+###############################
+
+When the provided support in Zephyr does not cover certain hardware platforms,
+some interfaces can be implemented directly by the platform or a custom
+architecture interface can be used and resulting in more flexibility and
```
Comment (aescolar) [inline doc/hardware/porting/custom.rst:8]: ".. can be used and resulting .."
This whole sentence is a bit too tough to read.

**Comment by nashif** [comment]
> the more I look into things and try to fix existing inconsistencies and interfaces, I realize how messy interrupt implementation is and how many ways exist to enable interrupt controllers that we need to fix ASAP, so for example, I do not think we need to support platform specific irq handling as this PR does and instead allow for this to be done using dedicated interrupt drivers, the same way we do this already on many architectures.

just seen this #66505

**Comment by ycsin** [comment]
@nashif any update for this PR?

**Pull request closed without merging**